### PR TITLE
Pipeline and task run eventing

### PR DESCRIPTION
**Pull request by steveodonovan**
# Changes

https://github.com/tektoncd/dashboard/issues/91

Moves the back end pipeline run websocket to /runs and includes the task run events.
Adds ties this into the store for realtime updates client side.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Comment by steveodonovan** [comment]
/hold

this needs further testing on an busy cluster.

**Comment by steveodonovan** [comment]
/assign @AlanGreene 

Verified on a busy cluster it doesn't fall over. Will add pipelines and task events over the same websocket in a seperate PR after this.

File: .eslintrc.js [hunk 1]
```diff
@@ -26,6 +26,7 @@ module.exports = {
       "react"
   ],
   "rules": {
+    "import/prefer-default-export": "off",
```
Comment (AlanGreene) [inline .eslintrc.js]: Yeah I see this rule as having limited value, I'm fine with disabling it. Can you also remove existing overrides in `src/constants/index.js` and `src/utils/test.js`?

File: src/containers/TaskRuns/TaskRuns.js [hunk 2]
```diff
@@ -108,14 +107,7 @@ export /* istanbul ignore next */ class TaskRunsContainer extends Component {
...
       };
     });
-    let notification;
-    if (taskRuns.length === 0) {
-      notification = {
-        kind: 'info',
-        message: 'Task has never run'
-      };
-    }
-    this.setState({ taskRuns, notification });
```
Comment (AlanGreene) [inline src/containers/TaskRuns/TaskRuns.js]: 👍

File: src/reducers/pipelineRuns.js [hunk 3]
```diff
@@ -19,6 +19,14 @@ import { ALL_NAMESPACES } from '../constants';
 
 function byId(state = {}, action) {
   switch (action.type) {
+    case 'PipelineRunCreated':
+    case 'PipelineRunUpdated':
+      const runById = { [action.payload.metadata.uid]: action.payload };
+      return merge({}, state, runById);
+    case 'PipelineRunDeleted':
+      const { ...newState } = state;
```
Comment (AlanGreene) [inline src/reducers/pipelineRuns.js]: this is typically written the other way around, unless extracting a specific field and lumping everything else together as `...rest`

`const newState = { ...state };` (read 'newState = shallow copy of state')

File: src/reducers/pipelineRuns.test.js [hunk 4]
```diff
@@ -53,6 +58,58 @@ it('PIPELINE_RUNS_FETCH_SUCCESS', () => {
   expect(selectors.isFetchingPipelineRuns(state)).toBe(false);
 });
 
+it('PipelineRunCreated', () => {
```
Comment (AlanGreene) [inline src/reducers/pipelineRuns.test.js]: 'PipelineRun events' maybe?

File: src/store/index.js [hunk 5]
```diff
@@ -15,20 +15,32 @@ import { applyMiddleware, compose, createStore } from 'redux';
...
     rootReducer,
     initialState,
     composeEnhancers(applyMiddleware(...middleware))
   );
+  return store;
 }
 
-const store = configureStore();
-export default store;
+export function getStore() {
```
Comment (AlanGreene) [inline src/store/index.js]: this function isn't really necessary, can just use `configureStore` which will return the existing store

File: src/index.js [hunk 6]
```diff
@@ -16,14 +16,22 @@ import ReactDOM from 'react-dom';
...
+import { getRunsWebSocketURL } from './api';
 
 import App from './containers/App';
 
+const runsWebSocket = new WebSocket(getRunsWebSocketURL());
+function closeSocket() {
+  runsWebSocket.close();
+}
+
+const store = configureStore({ runsWebSocket });
```
Comment (AlanGreene) [inline src/index.js]: should probably just call this `webSocket` since it'll handle more than runs

File: src/index.js [hunk 7]
```diff
@@ -16,14 +16,22 @@ import ReactDOM from 'react-dom';
...
+
 /* istanbul ignore next */
 ReactDOM.render(
   <Provider store={store}>
-    <App />
+    <App onUnload={closeSocket} />
```
Comment (AlanGreene) [inline src/index.js]: not seeing where this is used

File: src/index.js (hunk 6 above)
Comment (steveodonovan) [reply src/index.js]: mind if i tackle that in the next ? Going to change the api too

Comment (AlanGreene) [reply src/index.js]: works for me 👍

**Comment by steveodonovan** [comment]
@AlanGreene updated. Could I get another once over please ? Thanks

**Pull request merged**